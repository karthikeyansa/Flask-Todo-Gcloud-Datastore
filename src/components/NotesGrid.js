import React, { useEffect, useState } from "react";
import {API} from "../Apiservice";


function NotesGrid(props){
    const [notes, setNotes] = useState(props.Newnote);

    useEffect(()=>{
        async function notesChecker(){
            try{
                if(props.Newnote){
                    setNotes([props.Newnote, ...notes]);
                }
                else{
                    let response = await API.GetNotes();
                    if(response.success === true){
                        setNotes(response.result);
                    }
                }
            }
            catch(error){
                console.error(error);
            }
        }
        notesChecker();   
    }, [props.Newnote])

    const selectedNote = (note) => (event) =>{
        props.NoteSelected(note);
    } 

    const NoteRemover = (id) => (event) =>{
        async function deleteNote(){
            try{
                let response = await API.DeleteNote({id : id});
                if(response.success === true){
                    let displaycellRemoved = document.getElementById(id);
                    displaycellRemoved.style.display = "none";
                }
            }
            catch(error){
                console.error(error);
            }
        }
        deleteNote();
    }

    return(
        <React.Fragment>
            <div style={{display: "block",height: "100px"}}></div>
            <div className="displaygrid" id="displaygrid">
                {notes && notes.map((note)=>{
                    let completedTasks = 0;
                    let inCompletedTasks = 0;
                    return(
                    <div key={note.id}>
                        <div className="displaygridcell" id={note.id}>
                            <div style={{cursor: "pointer"}} onClick={selectedNote(note)}>
                                {note.content}
                            </div>
                                {note.tasks && note.tasks.forEach((task)=>{
                                    if (task.status === true){
                                        completedTasks += 1;
                                    }
                                    else if (task.status === false){
                                        inCompletedTasks += 1;
                                    }                     
                                })}
                            <div>
                                <span className="danger-btn" onClick={NoteRemover(note.id)}>üóëÔ∏è</span>&nbsp;
                                <span>Status: ‚ùåOngoing({inCompletedTasks}) <span style={{color: "lightgreen"}} >‚úî</span>Completed({completedTasks})</span>
                            </div>
                        </div>
                    </div>
                    )
                })}
            </div>
        </React.Fragment>
    )
}

export default NotesGrid;