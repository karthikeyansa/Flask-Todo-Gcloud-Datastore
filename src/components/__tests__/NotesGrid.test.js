import React from 'react';
import { mount, shallow } from 'enzyme';
import NotesGrid from '../NotesGrid';
import renderer from 'react-test-renderer';

describe('NotesGrid', () => {
  it('should render', () => {
    const wrapper = shallow(<NotesGrid />)
    expect(wrapper.exists()).toBeTruthy();
  })
})


describe('NotesGrid', () => {
  it('displaygrid rendering without crashing', () => {
    const wrapper = shallow(<NotesGrid />);
    expect(wrapper.find("#displaygrid").exists()).toBeTruthy();
  })

  it('displaygrid check data rendering with props old Data and new data', ()=>{
    const holder = document.createElement('div');
    document.body.appendChild(holder);

    const minprops = {
      allNotesData : [
        {id: 1, content: "Hello world", 
                tasks: [{task_id: 1, description: "this is task 1", status: true},
                        {task_id: 2, description: "this is task 2", status: false}]},
        {id: 2, content: "Hello mars", 
        tasks: [{task_id: 1, description: "go to mars", status: true},
                {task_id: 2, description: "drop bomb on the poles", status: false},
                {task_id: 3, description: "colonize mars", status: true}]}
      ],
      pushUpdatedNote : {id: 123, content: "sample hello world", 
                           tasks: [{task_id: 1, description: "this is task 1", status: true}, 
                                   {task_id: 2, description: "this is task 2", status: false}]},
      Newnote : null
    }

    const wrapper = mount(<NotesGrid {...minprops}/>, { attachTo: holder });
    expect(wrapper.find("div.displaygrid div.displaygridcell div#noteContent").exists()).toBeTruthy();
    const story1 = wrapper.find("div.displaygrid div.displaygridcell div#noteContent").first();
    expect(story1.text()).toBe(minprops.allNotesData[0].content);

    const storyOneDelete = wrapper.find("div.displaygrid div.displaygridcell span.danger-btn").first();
    expect(storyOneDelete.text()).toBe("🗑️");
    const storyOneTasksCount = wrapper.find("div.displaygrid div.displaygridcell span#statusHolder").first();
    expect(storyOneTasksCount.text()).toBe("Status: ❌Ongoing(1) ✔Completed(1)");

    const story2 = wrapper.find("div.displaygrid div.displaygridcell div#noteContent").at(1);
    expect(story2.text()).toBe(minprops.allNotesData[1].content);
    const storyTwoDelete = wrapper.find("div.displaygrid div.displaygridcell span.danger-btn").at(1);
    expect(storyTwoDelete.text()).toBe("🗑️");
    const storyTwoTasksCount = wrapper.find("div.displaygrid div.displaygridcell span#statusHolder").at(1);
    expect(storyTwoTasksCount.text()).toBe("Status: ❌Ongoing(1) ✔Completed(2)");

    const story2Remover = wrapper.find("div.displaygrid div.displaygridcell div span.danger-btn").first();
    const mockFetch = Promise.resolve({ json: () => Promise.resolve({success: true}) });
    global.fetch = jest.fn().mockImplementation(() => mockFetch);

    story2Remover.simulate('click')

    wrapper.unmount();
  })
})

describe('NotesGrid', () => {
  it('displaygrid check data rendering with props new data only', ()=>{
    const holder = document.createElement('div');
    document.body.appendChild(holder);

    const oneprops = {
      Newnote : [{id: 123, content: "sample one hello world", 
                           tasks: [{task_id: 1, description: "this is task 1", status: true}, 
                                   {task_id: 2, description: "this is task 2", status: false}]
      }],
    }

    const wrapper = mount(<NotesGrid {...oneprops}/>, { attachTo: holder });
    expect(wrapper.find("div.displaygrid div.displaygridcell div#noteContent").exists()).toBe(true);
  })
})

describe('NotesGrid', () => {
  it('displaygrid check data rendering with no props data', ()=>{
    const holder = document.createElement('div');
    document.body.appendChild(holder);
    
    const mockFetch = Promise.resolve({ json: () => Promise.resolve({success: true}) });
    global.fetch = jest.fn().mockImplementation(() => mockFetch);

    const wrapper = mount(<NotesGrid/>, { attachTo: holder });
    expect(wrapper.html()).toEqual(`<div class="displaygrid" id="displaygrid"></div>`);
  })
})

describe('NotesGrid', () => {
  it('should render without crashing as per in snapshot NotesGrid', () => {
      const tree = renderer.create(<NotesGrid />).toJSON();
      expect(tree).toMatchSnapshot();
  })
})