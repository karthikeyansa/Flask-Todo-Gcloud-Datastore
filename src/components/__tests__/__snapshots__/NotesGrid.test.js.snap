// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotesGrid should render without crashing as per in snapshot NotesGrid 1`] = `
<div
  className="displaygrid"
  id="displaygrid"
/>
`;
